src/Couchbase.Lite.Tests.Shared/ReplicationTest.cs

```diff
@@ -1275,7 +1280,225 @@ public void TestConflictResolverDeletedRemoteWin()
+                return doc;
+            });
+
+            RunReplication(config, 0, 0);
+
+            using (var db = Db.GetDocument("doc1")) {
+                db.GetString("wrong_id_key").Should().Be("wrong_id_value");
+            }
+        }
+        [Fact]
+        public void TestConflictResolverCalledTwice()
+        {
+            int resolveCnt = 0;
+            CreateReplicationConflict("doc1");
+
+            var config = CreateConfig(false, true, false);
+
+            config.ConflictResolver = new TestConflictResolver((conflict) =>
+            {
+                if (resolveCnt == 0) {
+                    using (var d = Db.GetDocument("doc1"))
+                    using (var doc = d.ToMutable()) {
+                        doc.SetString("name", "Cougar");
+                        Db.Save(doc);
+                    }
+                }
+
+                resolveCnt++;
+                return conflict.LocalDocument;
+            });
+
+            RunReplication(config, 0, 0);
+
+            resolveCnt.Should().Be(2);
+            using (var doc = Db.GetDocument("doc1")) {
+                    doc.GetString("name").Should().Be("Cougar");
+            }
+        }
+
+        [Fact]
+        public void TestNonBlockingDatabaseOperationConflictResolver()
+        {
+            int resolveCnt = 0;
+            CreateReplicationConflict("doc1");
+            var config = CreateConfig(false, true, false);
+            config.ConflictResolver = new TestConflictResolver((conflict) =>
+            {
+                if (resolveCnt == 0) {
+                    using (var d = Db.GetDocument("doc1"))
+                    using (var doc = d.ToMutable()) {
+                        d.GetString("name").Should().Be("Tiger");
+                        doc.SetString("name", "Cougar");
+                        Db.Save(doc);
+                        d.GetString("name").Should().Be("Cougar", "Because database save operation was not blocked");
+                    }
+                }
+                resolveCnt++;
+                return null;
+            });
+
+            RunReplication(config, 0, 0);
+
+            using (var doc = Db.GetDocument("doc1")) {
+                if(resolveCnt==1)
+                    doc.Should().BeNull();
+            }
+        }
+        
+        [Fact]
+        public void TestNonBlockingConflictResolver()
+        {
+            CreateReplicationConflict("doc1");
+            CreateReplicationConflict("doc2");
+            var config = CreateConfig(false, true, false);
+            Queue<string> q = new Queue<string>();
+            var wa = new WaitAssert();
+            config.ConflictResolver = new TestConflictResolver((conflict) =>
+            {
+                var cnt = 0;
+                lock (q) {
+                    q.Enqueue(conflict.LocalDocument.Id);
+                    cnt = q.Count;
+                }
+
+                if (cnt == 1) {
+                    Thread.Sleep(500);
```

Comment:
This is kinda flakey.  Does iOS have latches of some kind, so that you can wait for thee other guy to finish, instead of waiting for time?

Reply:
Yeah, that's a valid improvement.  Here is how it could play out:

`ManualResetEvent` outside of the callback.  If `cnt` is 1 then block on it, otherwise ignore it.  Do the rest of the logic, then if `cnt` is *not* 1 signal the event.